go-controller/pkg/ovn/ovn.go

```diff
@@ -716,23 +599,34 @@ func (oc *Controller) WatchPods() {
 					podNs, podName)
 				return
 			}
-			if !oc.ensurePod(oldPod, pod, oc.checkAndSkipRetryPod(pod.UID)) {
+			if err := oc.ensurePod(oldPod, pod, oc.checkAndSkipRetryPod(pod)); err != nil {
 				// unskip failed pod for next retry iteration
 				oc.unSkipRetryPod(pod)
 				return
 			}
-			oc.checkAndDeleteRetryPod(pod.UID)
+			oc.checkAndDeleteRetryPod(pod)
 		},
 		DeleteFunc: func(obj interface{}) {
 			pod := obj.(*kapi.Pod)
 			go oc.metricsRecorder.CleanPodRecord(pod.UID)
-			oc.checkAndDeleteRetryPod(pod.UID)
 			if !util.PodWantsNetwork(pod) {
 				oc.deletePodExternalGW(pod)
+				oc.checkAndDeleteRetryPod(pod) // no retries for host-networked pods
 				return
 			}
+			oc.initRetryDelPod(pod)
+			// we have a copy of portInfo in the retry cache now, we can remove it from
+			// logicalPortCache so that we don't race with a new add pod that comes with
+			// the same name
+			oc.logicalPortCache.remove(util.GetLogicalPortName(pod.Namespace, pod.Name))
 			// deleteLogicalPort will take care of removing exgw for ovn networked pods
-			oc.deleteLogicalPort(pod)
+			if err := oc.deleteLogicalPort(pod); err != nil {
+				oc.recordPodEvent(err, pod)
```

Comment:
we record pod event here in the watcher, but on add we do it inside of ensurePod everywhere, maybe for add we should just do it outside of the ensure function as well?